src/p4p/test/system_client_server.py

```diff
@@ -0,0 +1,631 @@
+                ('return', ('S', 'malcolm:core/StringMeta:1.0', [
+                    ('description', 's'),
+                    ('tags', 'as'),
+                    ('writeable', '?'),
+                    ('label', 's')
+                ])),
+            ])),
+            ('required', 'as')
+        ])),
+    ])),
+    ('error', empty_method_tuple)
+], 'malcolm:core/Block:1.0')
+
+hello_dict = {
+    'meta': {
+        'description': 'Hardware Block with a greet() Method',
+        'tags': [],
+        'writeable': False,
+        'label': 'TESTHELLO',
+        'fields': ['health', 'greet', 'error']
+    },
+    'health': {
+        'value': "OK",
+        'alarm': alarm_ok,
+        'timeStamp': ts_zero,
+        'meta': {
+            'description': 'Displays OK or an error message',
+            'tags': ['widget:textupdate'],
+            'writeable': False,
+            'label': 'Health'
+        }
+    },
+    'greet': {
+        'takes': {
+            'elements': {
+                'name': {
+                    'description': 'The name of the person to greet',
+                    'tags': ['widget:textinput'],
+                    'writeable': True,
+                    'label': 'Name'
+                },
+                'sleep': {
+                    'dtype': 'float64',
+                    'description': 'Time to wait before returning',
+                    'tags': ['widget:textinput'],
+                    'writeable': True,
+                    'label': 'Sleep'
+                }
+            },
+            'required': ["name"]
+        },
+        'defaults': {
+            'sleep': 0.0
+        },
+        'description': 'Optionally sleep <sleep> seconds, then return a greeting to <name>',
+        'tags': [],
+        'writeable': True,
+        'label': 'Greet',
+        'returns': {
+            'elements': {
+                'return': {
+                    'description': 'The manufactured greeting',
+                    'tags': ['widget:textupdate'],
+                    'writeable': False,
+                    'label': 'Return'
+                }
+            },
+            'required': ['return']
+        }
+    },
+    'error': {
+        'takes': empty_map_meta_dict,
+        'description': 'Raise an error',
+        'tags': [],
+        'writeable': True,
+        'label': 'Error',
+        'returns': empty_map_meta_dict
+    }
+}
+
+hello_expected = Value(hello_block_t, hello_dict)
+
+
+class Channel(object):
+    # TODO: Why is this needed?
+    Value = Value
+
+    def __init__(self, structure, field_name=None):
+        self.structure = structure
+        self.field_name = field_name
+
+    def get(self):
+        if self.field_name:
+            ret = self.structure[self.field_name]
+        else:
+            ret = self.structure
+        return ret
+
+    def monitor(self, updater):
+        # TODO: what is good here?
+        self.structure.attach(updater, self.field_name)
```

Comment:
Actually, for my use case, all I want is an updater function that I can call when I have changed the structure, telling it what bitset has changed. I think I will retract this particular idea, because attaching an updater to the structure so it is automatically called is probably a bit too limited (how do you do updates of a number of fields at once to appear in a single changeset otherwise?)

Reply:
Good.  I'd like to specify a higher level API(s) at this point.  As with the client API, I'm planning to have a lower-level API at the boundary between C++ and python code.  However, I'm not going to figure out the best form for this API until I start writing it.

Reply:
cf. https://github.com/mdavidsaver/p4p/wiki/serverapi